### ISSUE TITLE
Feature request: support for split/Subset-like behavior and shared local cache in StreamingDataset

### ISSUE THREAD

**Issue by Nima-Hs** [mosaicml/streaming#916]
<html>
<body>
<html><head></head><body>
<p>Hi, and thank you for the excellent work on this package!</p>
<p>I've been using <code inline="">StreamingDataset</code> for large-scale training with a dataset stored on S3, and I've encountered two major limitations that I couldn't find addressed in the documentation. I’d appreciate clarification or guidance on best practices, and I'd like to propose feature support if these aren't currently handled.</p>
<hr>
<h3>📦 Use Case</h3>
<p>We’re working with a very large dataset (~100 TB) stored in an S3-compatible object storage system. The dataset is preprocessed into MDS format and hosted in a shared S3 bucket. Each sample includes metadata (e.g., <code inline="">age</code>, <code inline="">sex</code>, etc.), and users often want to:</p>
<ol>
<li>
<p><strong>Filter the dataset</strong> based on metadata (e.g., only use female patients over age 60)</p>
</li>
<li>
<p><strong>Split the dataset</strong> into train/val/test subsets using arbitrary logic</p>
</li>
</ol>
<p>This is a common scenario in medical and scientific domains where researchers need flexible subsets from a shared, centralized dataset.</p>
<hr>
<h3>❗ Current Limitation 1: Subset behavior unclear and incompatible with DDP</h3>
<p>For <strong>single-GPU training</strong>, we’ve been able to use <code inline="">torch.utils.data.Subset(streaming_dataset, indices)</code> to split or filter the dataset, and it seems to work as expected.</p>
<p>However, this usage is <strong>not documented anywhere</strong>, and I want to confirm:</p>
<ul>
<li>
<p>Is using <code inline="">Subset</code> on a <code inline="">StreamingDataset</code> officially supported?</p>
</li>
<li>
<p>Are there caveats around <code inline="">__len__</code>, shuffling, or streaming behavior when using <code inline="">Subset</code> this way?</p>
</li>
<li>
<p>Will it break prefetching or caching mechanisms internally?</p>
</li>
</ul>
<p>For <strong>multi-GPU training (DDP)</strong>, the problem becomes more serious:<br>
As soon as the <code inline="">StreamingDataset</code> is initialized, it <strong>automatically shards the dataset</strong> across processes based on <code inline="">RANK</code>, <code inline="">WORLD_SIZE</code>, etc. This means:</p>
<ul>
<li>
<p>Any filtering or splitting logic (e.g., <code inline="">Subset</code>) only applies to that local shard</p>
</li>
<li>
<p>The user has no way to enforce a consistent global split across ranks</p>
</li>
<li>
<p>It becomes impossible to implement deterministic train/val/test splits or metadata-based filtering unless the dataset is physically reprocessed into new shard directories (which is infeasible at 100 TB scale)</p>
</li>
</ul>
<p>We would benefit greatly from:</p>
<ul>
<li>
<p>A mechanism to <strong>restrict the global sample list before DDP sharding occurs</strong>, such as a <code inline="">split()</code> method or <code inline="">partition_index</code> constructor argument that defines a subset of the dataset globally</p>
</li>
<li>
<p>Or a way to <strong>defer sharding</strong> so that the user can apply their own filtering or sampling logic prior to partitioning for distributed training</p>
</li>
</ul>
<hr>
<h3>❗ Current Limitation 2: Shared local cache not supported</h3>
<p>In our cluster setup, multiple users and jobs need to train on the same dataset. To save bandwidth and storage, we want to <strong>share the same local cache directory</strong> (e.g., <code inline="">/scratch/mds_cache</code>) between users or jobs.</p>
<p>Currently:</p>
<ul>
<li>
<p>Each process or user creates its own local cache path, leading to duplicated downloads</p>
</li>
<li>
<p>Reusing the same path leads to potential race conditions, errors, or inconsistent states</p>
</li>
</ul>
<p>We would love to see support for:</p>
<ul>
<li>
<p>A <strong>safe shared cache mode</strong> — possibly with options for:</p>
<ul>
<li>
<p>Read-only access</p>
</li>
<li>
<p>Locking mechanisms</p>
</li>
<li>
<p>Checksum validation for partially downloaded files</p>
</li>
</ul>
</li>
</ul>
<p>This would make streaming much more scalable for large shared environments.</p>
</body>
</html>
